src/py/mat3ra/made/tools/build/defect/builders.py

```diff
@@ -371,3 +378,143 @@ def _generate(self, configuration: _ConfigurationType) -> List[_GeneratedItemTyp
+        normalized_direction_vector = direction_vector / np.linalg.norm(direction_vector)
+        return normalized_direction_vector
+
+    def _calcualte_height(self, material, new_material):
+        original_max_z = get_atomic_coordinates_extremum(material, use_cartesian_coordinates=True)
+        added_layers_max_z = get_atomic_coordinates_extremum(new_material, use_cartesian_coordinates=True)
+        height_cart = added_layers_max_z - original_max_z
+        return height_cart
+
+    def _calculate_rotation_parameters(self, original_material, new_material, normalized_direction_vector):
+        """Calculate the necessary rotation angle and axis."""
+        height_cart = self._calcualte_height(original_material, new_material)
+        cut_direction_xy_proj_cart = np.linalg.norm(
+            np.dot(np.array(new_material.basis.cell.vectors_as_nested_array), normalized_direction_vector)
+        )
+        # Slope of the terrace along the cut direction
+        hypotenuse = np.linalg.norm([height_cart, cut_direction_xy_proj_cart])
+        angle = np.arctan(height_cart / cut_direction_xy_proj_cart) * 180 / np.pi
+        normalized_rotation_axis = np.cross(normalized_direction_vector, [0, 0, 1]).tolist()
+        delta_length = hypotenuse - cut_direction_xy_proj_cart
+        return angle, normalized_rotation_axis, delta_length
+
+    def _increase_lattice_size(self, material, length_increase, direction_of_increase):
+        """Adjust the lattice vectors to accommodate the new structure dimensions."""
+        vector_a, vector_b = np.array(material.basis.cell.vector1), np.array(material.basis.cell.vector2)
+        delta_a_cart = np.dot(vector_a, direction_of_increase) * length_increase / np.linalg.norm(vector_a)
+        delta_b_cart = np.dot(vector_b, direction_of_increase) * length_increase / np.linalg.norm(vector_b)
+        vector_a_prime = vector_a + delta_a_cart * vector_a / np.linalg.norm(vector_a)
+        vector_b_prime = vector_b + delta_b_cart * vector_b / np.linalg.norm(vector_b)
+
+        cart_basis = material.basis.copy()
+        cart_basis.to_cartesian()
+        cart_basis.cell.vector1 = vector_a_prime
+        cart_basis.cell.vector2 = vector_b_prime
+        material.basis = cart_basis
+
+        new_lattice = material.lattice.clone()
+        new_lattice.a = new_lattice.a + delta_a_cart
+        new_lattice.b = new_lattice.b + delta_b_cart
+        material.lattice = new_lattice
+        return material
+
+    def _update_material_name(self, material, configuration):
+        updated_material = super()._update_material_name(material, configuration)
+        new_name = f"{updated_material.name}, {configuration.steps_number}-step Terrace {configuration.cut_direction}"
+        updated_material.name = new_name
+        return updated_material
+
+    def _rotate_material(self, material: Material, axis: List[int], angle: float) -> Material:
```

Comment:
This is a generic function and can go to `modify` for material tools

Reply:
It means I need to add test for this too